### PR TITLE
T2719: prototype of an op mode command runner based on type hints and introspection

### PR DESCRIPTION
**Pull request by dmbaturin**
## Change Summary

This is a prototype of a fully automated op mode runner that eliminates a need to write option parsing or argument passing by hand for a large class of commands.

It generates script subcommands from functions with standardized names (`show_*`, `reset_*`...) and also generates subcommand option parsers from type annotations of those functions.

Then it automatically calls specific functions determined from subcommands. For example, `neighbor.py show --family inet --interface eth0` automatically calls `show(raw=False, family="inet", interface="eth0", state=None`.

The end goal is to allow op mode script authors to simply add functions with correct names and type annotations rather than handle calling those functions by hand. Then we can automatically generate GraphQL schemas from them, and _possibly_ even XML definitions (or at least base files for them that can be improved by hand-editing).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T2719

## Component(s) name

Op mode.

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Review by jestabro — APPROVED** [review]

**Review by jestabro — APPROVED** [review]

File: python/vyos/opmode.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                    subparser.add_argument(f"--{opt}", type=_get_arg_type(th), required=True)
+
+    # Get options as a dict rather than a namespace,
+    # so that we can modify it and pack for passing to functions
+    args = vars(parser.parse_args())
+
+    if not args["subcommand"]:
+       print("Subcommand required!")
+       parser.print_usage()
+       sys.exit(1)
```
Comment (sever-sever) [inline python/vyos/opmode.py]: Could you check it? 
It seems only 3 whitespaces

File: src/op_mode/cpu.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+import json
+import sys
+import typing
+
+import vyos.cpu
+import vyos.opmode
+
+from jinja2 import Template
+from vyos.util import popen, DEVNULL
```
Comment (sever-sever) [inline src/op_mode/cpu.py]: `json, typing and popen` not used in the script script

File: src/op_mode/version.py [hunk 3]
```diff
@@ -19,11 +19,14 @@
 #    Used by the "run show version" command.
 
 import argparse
```
Comment (sever-sever) [inline src/op_mode/version.py]: Not used

File: src/op_mode/version.py [hunk 4]
```diff
@@ -19,11 +19,14 @@
...
+import sys
+import typing
+
+import vyos.opmode
 import vyos.version
 import vyos.limericks
 
 from jinja2 import Template
-from sys import exit
 from vyos.util import call
```
Comment (sever-sever) [inline src/op_mode/version.py]: `call` not used in the script

**Review by sever-sever — CHANGES_REQUESTED** [review]

**Review by sever-sever — APPROVED** [review]

**Pull request merged**